pkg/collector/corechecks/servicediscovery/module/envs.go

```diff
@@ -61,55 +44,103 @@ func readEnvsFile(path string) ([]string, error) {
 // lrwx------ 1 foo foo 64 Aug 13 14:24 2 -> /dev/pts/6
 // lrwx------ 1 foo foo 64 Aug 13 14:24 3 -> '/memfd:dd_environ (deleted)'
 // ```
-func getInjectedEnvs(proc *process.Process) []string {
-	fdsPath := kernel.HostProc(strconv.Itoa(int(proc.Pid)), "fd")
-	entries, err := os.ReadDir(fdsPath)
+func getInjectionMeta(proc *process.Process) (*InjectedProcess, bool) {
+	path, found := findInjectorFile(proc)
+	if !found {
+		return nil, false
+	}
+	injectionMeta, err := extractInjectionMeta(path)
 	if err != nil {
-		return nil
+		log.Warnf("error extracting injected envs: %s", err)
+		return nil, false
 	}
+	return injectionMeta, true
 
-	for _, entry := range entries {
-		path := filepath.Join(fdsPath, entry.Name())
-		name, err := os.Readlink(path)
-		if err != nil {
-			continue
-		}
+}
 
-		if name != injectorMemFdPath {
-			continue
-		}
+func extractInjectionMeta(path string) (*InjectedProcess, error) {
+	reader, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer reader.Close()
 
-		envs, _ := readEnvsFile(path)
-		return envs
+	data, err := io.ReadAll(io.LimitReader(reader, memFdMaxSize))
+	if err != nil {
+		return nil, err
+	}
+	if len(data) == memFdMaxSize {
+		return nil, io.ErrShortBuffer
 	}
 
-	return nil
+	var injectedProc InjectedProcess
+	_, err = injectedProc.UnmarshalMsg(data)
+	if err != nil {
+		return nil, err
+	}
+	return &injectedProc, nil
 }
 
-// envsToMap splits a list of strings containing environment variables of the
-// format NAME=VAL to a map.
-func envsToMap(envs ...string) map[string]string {
-	envMap := make(map[string]string, len(envs))
-	for _, env := range envs {
-		name, val, found := strings.Cut(env, "=")
-		if !found {
+func findInjectorFile(proc *process.Process) (string, bool) {
+	fdsPath := kernel.HostProc(strconv.Itoa(int(proc.Pid)), "fd")
+	// quick path, the shadow file is the first opened file by the process
+	// unless there are inherited fds
+	path := filepath.Join(fdsPath, "3")
+	if isInjectorFile(path) {
+		return path, true
+	}
+	entries, err := os.ReadDir(fdsPath)
+	if err != nil {
+		return "", false
+	}
+	for _, entry := range entries {
+		switch entry.Name() {
+		case "0", "1", "2", "3":
 			continue
+		default:
+			path := filepath.Join(fdsPath, entry.Name())
+			if isInjectorFile(path) {
+				return path, true
+			}
 		}
+	}
```

Comment:
Since you care only from the names, you should use `os.Readdirnames(-1)`

Reply:
nice, if this becomes expensive we can easily limit the number of fd checked per process. Given that it's always the first file opened by the process, it should be at the top.
I don't know how frequently processes pass their fd to child ones